src/include/platform/internal/simulation/PlatformManagerImpl.cpp

```diff
@@ -0,0 +1,438 @@
+/*
+ *
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *          Provides an generic implementation of PlatformManager features
+ *          for use on Simulated platforms.
+ */
+
+#pragma once
+
+#include <platform/DeviceControlServer.h>
+#include <platform/PlatformManager.h>
+#include <platform/internal/CHIPDeviceLayerInternal.h>
+
+#include <signal.h>
+#include <unistd.h>
+
+using namespace ::chip::app::Clusters;
+
+namespace chip {
+namespace DeviceLayer {
+
+namespace {
+void SignalHandler(int signum)
```

Comment:
This file likely should not be in platform/internal. The way this is organized right now, it will force that all "simulated devices" have same behavior by platform linkage.

This particular impl here basically makes it impossible to ever refactor PlatformManager to be more easily portable/replaceable (similar to what was done with ConfigurationManager) because all the singleton behavior will now be a dependence of passing cert tests, but glued by this fake "platform" impl. Nobody will ever want to touch PlatfomManager at risk of breaking a variety of cert tests that have expectations baked here.

We need to find something much more scalable. 

At the very least, this implementation should delegate to a pluggable/injectable "behavior" module for the PlatformManagerImpl so that the exact simulated device behavior is not truly/fully limited by this one gigantic hardcoded version.

Reply:
> This particular impl here basically makes it impossible to ever refactor PlatformManager to be more easily portable/replaceable (similar to what was done with ConfigurationManager) because all the singleton behavior will now be a dependence of passing cert tests, but glued by this fake "platform" impl. Nobody will ever want to touch PlatfomManager at risk of breaking a variety of cert tests that have expectations baked here.

I'm not fully agreeing on that but that's OK because I didn't like what I did too.

> We need to find something much more scalable.

Fair enough. In the meantime I need to move forward so I revert the change in a way where the code is duplicated between Linux and Mac again.